tests/components/homeassistant_sky_connect/test_init.py

```diff
@@ -58,3 +88,129 @@ async def test_config_entry_migration_v2(hass: HomeAssistant) -> None:
+
+
+async def test_usb_device_reactivity(
+    hass: HomeAssistant, hass_ws_client: WebSocketGenerator
+) -> None:
+    """Test setting up USB monitoring."""
+
+    # Disable the inotify subsystem to allow us to trigger rescans manually
+    with patch(
+        "homeassistant.components.usb.AIOUSBWatcher",
+        side_effect=InotifyNotAvailableError,
+    ):
+        assert await async_setup_component(hass, "usb", {"usb": {}})
+
+    await hass.async_block_till_done()
+    hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
+    await hass.async_block_till_done()
+    ws_client = await hass_ws_client(hass)
+
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        unique_id="some_unique_id",
+        data={
+            "description": "SkyConnect v1.0",
+            "device": "/dev/serial/by-id/usb-Nabu_Casa_SkyConnect_v1.0_9e2adbd75b8beb119fe564a0f320645d-if00-port0",
+            "vid": "10C4",
+            "pid": "EA60",
+            "serial_number": "3c0ed67c628beb11b1cd64a0f320645d",
+            "manufacturer": "Nabu Casa",
+            "product": "SkyConnect v1.0",
+            "firmware": "ezsp",
+            "firmware_version": "7.4.4.0",
+        },
+        version=1,
+        minor_version=3,
+    )
+
+    config_entry.add_to_hass(hass)
+
+    with patch(
+        "homeassistant.components.homeassistant_sky_connect.os.path.exists"
+    ) as mock_exists:
+        mock_exists.return_value = False
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+        # Failed to set up, the device is missing
+        assert config_entry.state == ConfigEntryState.SETUP_RETRY
+
+        # Now we make it available but do not wait
+        mock_exists.return_value = True
+
+        with patch(
+            "homeassistant.components.usb.comports",
+            return_value=[
+                create_list_port_info(
+                    device="/dev/serial/by-id/usb-Nabu_Casa_SkyConnect_v1.0_9e2adbd75b8beb119fe564a0f320645d-if00-port0",
+                    vid=0x10C4,
+                    pid=0xEA60,
+                    serial_number="3c0ed67c628beb11b1cd64a0f320645d",
+                    manufacturer="Nabu Casa",
+                    description="SkyConnect v1.0",
+                )
+            ],
+        ):
+            await async_request_scan(hass, ws_client, req_id=1)
+
+        # It loads immediately
+        await hass.async_block_till_done(wait_background_tasks=True)
+        await hass.async_block_till_done(wait_background_tasks=True)
+        assert config_entry.state == ConfigEntryState.LOADED
+
+        # Wait for a bit for the USB scan debouncer to cool off
+        async_fire_time_changed(hass, dt_util.now() + timedelta(minutes=5))
+
+        # Unplug the stick
+        mock_exists.return_value = False
+
+        with patch("homeassistant.components.usb.comports", return_value=[]):
+            await async_request_scan(hass, ws_client, req_id=2)
```

Comment:
Can we avoid patching the internals of `usb` and patch a public API instead?

Reply:
Check out my second PR that should facilitate this: <https://github.com/home-assistant/core/pull/141521>. With it, you would patch `usb.scan_serial_ports` instead of `usb.comports`.

Other than this change, I'm not too sure how else to mock this API. Integrations are pushed serial port information by the USB integration.

Reply:
Here's a slightly smaller PR: https://github.com/home-assistant/core/pull/141733